File: central_ios/BlerpcCentral/Client/BlerpcClient.swift
import BlerpcProtocol
import CoreBluetooth
import Foundation
import os.log

enum BlerpcClientError: Error {
    case notConnected
    case payloadTooLarge(actual: Int, limit: Int)
    case responseTooLarge
    case peripheralError(code: UInt8)
    case unexpectedResponseType(UInt8)
    case commandNameMismatch(expected: String, got: String)
    case keyExchangeFailed(String)
    case encryptionError(String)
    case replayDetected(counter: UInt32)
}

private let logger = Logger(subsystem: "com.blerpc", category: "BlerpcClient")

final class BlerpcClient: GeneratedClientProtocol {
    let transport = BleTransport()
    private var splitter: ContainerSplitter?
    private let assembler = ContainerAssembler()
    private var timeoutMs: Int = 100
    private var maxRequestPayloadSize: Int?
    private var maxResponsePayloadSize: Int?

    private var session: BlerpcCryptoSession?

    var mtu: Int { transport.mtu }
    var isEncrypted: Bool { session != nil }

    func scan(
        timeout: TimeInterval = 5,
        serviceUUID filterUUID: CBUUID? = serviceUUID
    ) async throws -> [ScannedDevice] {
        return try await transport.scan(timeout: timeout, serviceUUID: filterUUID)
    }

    func connect(device: ScannedDevice) async throws {
        try await transport.connect(device: device)
        let mtuVal = transport.mtu
        splitter = ContainerSplitter(mtu: mtuVal)

        do {
            try await requestTimeout()
        } catch is BleTransportError {
            logger.debug("Peripheral did not respond to timeout request, using default")
        }
        do {
            try await requestCapabilities()
        } catch is BleTransportError {
            logger.debug("Peripheral did not respond to capabilities request")
        }
    }

    private func requestTimeout() async throws {
        guard let s = splitter else { throw BlerpcClientError.notConnected }
        let tid = s.nextTransactionId()
        let req = makeTimeoutRequest(transactionId: tid)
        try transport.write(req.serialize())
        let data = try await transport.readNotify(timeoutMs: 1000)
        let resp = try Container.deserialize(data)
        if resp.containerType == .control,
           resp.controlCmd == .timeout,
           resp.payload.count == 2 {
            let ms = Int(resp.payload[0]) | (Int(resp.payload[1]) << 8)
            timeoutMs = ms
        } else {
            logger.warning("Unexpected timeout response: type=\(resp.containerType.rawValue), cmd=\(resp.controlCmd.rawValue), payload_len=\(resp.payload.count)")
        }
    }

    private func requestCapabilities() async throws {
        guard let s = splitter else { throw BlerpcClientError.notConnected }
        let tid = s.nextTransactionId()
        let req = makeCapabilitiesRequest(transactionId: tid)
        try transport.write(req.serialize())
        let data = try await transport.readNotify(timeoutMs: 1000)
        let resp = try Container.deserialize(data)
        if resp.containerType == .control,
           resp.controlCmd == .capabilities,
           resp.payload.count >= 6 {
            let maxReq = Int(resp.payload[0]) | (Int(resp.payload[1]) << 8)
            let maxResp = Int(resp.payload[2]) | (Int(resp.payload[3]) << 8)
            let flags = Int(resp.payload[4]) | (Int(resp.payload[5]) << 8)
            if maxReq == 0 || maxResp == 0 {
                logger.warning("Peripheral reported zero capability: max_request=\(maxReq), max_response=\(maxResp)")
            }
            maxRequestPayloadSize = maxReq
            maxResponsePayloadSize = maxResp
            logger.info("Peripheral capabilities: max_request=\(maxReq), max_response=\(maxResp), flags=0x\(String(flags, radix: 16, uppercase: false))")

            if flags & Int(capabilityFlagEncryptionSupported) != 0 {
                try await performKeyExchange()
            }
        } else {
            logger.warning("Unexpected capabilities response: type=\(resp.containerType.rawValue), cmd=\(resp.controlCmd.rawValue), payload_len=\(resp.payload.count)")
        }
    }

    private func performKeyExchange() async throws {
        guard let s = splitter else { throw BlerpcClientError.notConnected }

        session = try await BlerpcCrypto.centralPerformKeyExchange(
            send: { payload in
                let tid = s.nextTransactionId()
                let req = makeKeyExchange(transactionId: tid, payload: payload)
                try self.transport.write(req.serialize())
            },
            receive: {
                let data = try await self.transport.readNotify(timeoutMs: 2000)
                let resp = try Container.deserialize(data)
                guard resp.containerType == .control,
                      resp.controlCmd == .keyExchange else {
                    throw BlerpcClientError.keyExchangeFailed("Expected KEY_EXCHANGE response")
                }
                return resp.payload
            }
        )
        logger.info("E2E encryption established")
    }

    private func encryptPayload(_ payload: Data) throws -> Data {
        guard let s = session else { return payload }
        return try s.encrypt(payload)
    }

    private func decryptPayload(_ payload: Data) throws -> Data {
        guard let s = session else { return payload }
        return try s.decrypt(payload)
    }

    func call(cmdName: String, requestData: Data) async throws -> Data {
        guard let s = splitter else { throw BlerpcClientError.notConnected }

        let cmd = CommandPacket(cmdType: .request, cmdName: cmdName, data: requestData)
        let payload = try cmd.serialize()

        if let limit = maxRequestPayloadSize, payload.count > limit {
            throw BlerpcClientError.payloadTooLarge(actual: payload.count, limit: limit)
        }

        let sendPayload = try encryptPayload(payload)
        let containers = try s.split(sendPayload)
        for c in containers {
            try transport.write(c.serialize())
        }

        assembler.reset()
        var firstRead = true
        while true {
            let t = firstRead ? max(timeoutMs, 2000) : timeoutMs
            firstRead = false
            let notifyData = try await transport.readNotify(timeoutMs: t)
            let container = try Container.deserialize(notifyData)

            if container.containerType == .control {
                if container.controlCmd == .error, !container.payload.isEmpty {
                    let errorCode = container.payload[0]
                    if errorCode == blerpcErrorResponseTooLarge {
                        throw BlerpcClientError.responseTooLarge
                    }
                    throw BlerpcClientError.peripheralError(code: errorCode)
                }
                continue
            }

            if let result = assembler.feed(container) {
                let decrypted = try decryptPayload(result)
                let resp = try CommandPacket.deserialize(decrypted)
                guard resp.cmdType == .response else {
                    throw BlerpcClientError.unexpectedResponseType(resp.cmdType.rawValue)
                }
                guard resp.cmdName == cmdName else {
                    throw BlerpcClientError.commandNameMismatch(
                        expected: cmdName, got: resp.cmdName
                    )
                }
                return resp.data
            }
        }
    }

    func streamReceive(cmdName: String, requestData: Data) async throws -> [Data] {
        guard let s = splitter else { throw BlerpcClientError.notConnected }

        let cmd = CommandPacket(cmdType: .request, cmdName: cmdName, data: requestData)
        let payload = try cmd.serialize()

        if let limit = maxRequestPayloadSize, payload.count > limit {
            throw BlerpcClientError.payloadTooLarge(actual: payload.count, limit: limit)
        }

        let sendPayload = try encryptPayload(payload)
        let containers = try s.split(sendPayload)
        for c in containers {
            try transport.write(c.serialize())
        }

        var results: [Data] = []
        assembler.reset()
        var firstRead = true
        while true {
            let t = firstRead ? max(timeoutMs, 2000) : timeoutMs
            firstRead = false
            let notifyData = try await transport.readNotify(timeoutMs: t)
            let container = try Container.deserialize(notifyData)

            if container.containerType == .control {
                if container.controlCmd == .streamEndP2C {
                    break
                }
                if container.controlCmd == .error, !container.payload.isEmpty {
                    let errorCode = container.payload[0]
                    if errorCode == blerpcErrorResponseTooLarge {
                        throw BlerpcClientError.responseTooLarge
                    }
                    throw BlerpcClientError.peripheralError(code: errorCode)
                }
                continue
            }

            if let result = assembler.feed(container) {
                let decrypted = try decryptPayload(result)
                let resp = try CommandPacket.deserialize(decrypted)
                guard resp.cmdType == .response else {
                    throw BlerpcClientError.unexpectedResponseType(resp.cmdType.rawValue)
                }
                results.append(resp.data)
            }
        }
        return results
    }

    func streamSend(
        cmdName: String,
        messages: [Data],
        finalCmdName: String
    ) async throws -> Data {
        guard let s = splitter else { throw BlerpcClientError.notConnected }

        for msgData in messages {
            let cmd = CommandPacket(cmdType: .request, cmdName: cmdName, data: msgData)
            let payload = try cmd.serialize()
            let sendPayload = try encryptPayload(payload)
            let containers = try s.split(sendPayload)
            for c in containers {
                try transport.write(c.serialize())
            }
        }

        // Send STREAM_END_C2P
        let tid = s.nextTransactionId()
        let streamEnd = makeStreamEndC2P(transactionId: tid)
        try transport.write(streamEnd.serialize())

        // Wait for final response
        assembler.reset()
        var firstRead = true
        while true {
            let t = firstRead ? max(timeoutMs, 2000) : timeoutMs
            firstRead = false
            let notifyData = try await transport.readNotify(timeoutMs: t)
            let container = try Container.deserialize(notifyData)

            if container.containerType == .control {
                if container.controlCmd == .error, !container.payload.isEmpty {
                    let errorCode = container.payload[0]
                    if errorCode == blerpcErrorResponseTooLarge {
                        throw BlerpcClientError.responseTooLarge
                    }
                    throw BlerpcClientError.peripheralError(code: errorCode)
                }
                continue
            }

            if let result = assembler.feed(container) {
                let decrypted = try decryptPayload(result)
                let resp = try CommandPacket.deserialize(decrypted)
                guard resp.cmdType == .response else {
                    throw BlerpcClientError.unexpectedResponseType(resp.cmdType.rawValue)
                }
                guard resp.cmdName == finalCmdName else {
                    throw BlerpcClientError.commandNameMismatch(
                        expected: finalCmdName, got: resp.cmdName
                    )
                }
                return resp.data
            }
        }
    }

    func counterStreamAll(count: UInt32) async throws -> [(seq: UInt32, value: Int32)] {
        var req = Blerpc_CounterStreamRequest()
        req.count = count
        let responses = try await streamReceive(
            cmdName: "counter_stream",
            requestData: try req.serializedData()
        )
        return try responses.map { data in
            let resp = try Blerpc_CounterStreamResponse(serializedBytes: data)
            return (seq: resp.seq, value: resp.value)
        }
    }

    func counterUploadAll(count: Int) async throws -> Blerpc_CounterUploadResponse {
        let messages = try (0..<count).map { i -> Data in
            var req = Blerpc_CounterUploadRequest()
            req.seq = UInt32(i)
            req.value = Int32(i * 10)
            return try req.serializedData()
        }
        let respData = try await streamSend(
            cmdName: "counter_upload",
            messages: messages,
            finalCmdName: "counter_upload"
        )
        return try Blerpc_CounterUploadResponse(serializedBytes: respData)
    }

    func disconnect() {
        transport.disconnect()
    }
}

File: central_android/app/src/main/java/com/blerpc/android/client/BlerpcClient.kt
package com.blerpc.android.client

import android.content.Context
import android.util.Log
import com.blerpc.android.ble.BleTransport
import com.blerpc.android.ble.SERVICE_UUID
import com.blerpc.android.ble.ScannedDevice
import com.blerpc.protocol.*
import com.blerpc.protocol.BlerpcCryptoSession
import com.blerpc.protocol.CentralKeyExchange
import com.blerpc.protocol.CAPABILITY_FLAG_ENCRYPTION_SUPPORTED
import com.blerpc.protocol.makeKeyExchange
import kotlinx.coroutines.TimeoutCancellationException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

class PayloadTooLargeError(actual: Int, limit: Int) :
    Exception("Request payload ($actual bytes) exceeds peripheral limit ($limit bytes)")

class ResponseTooLargeError(message: String) : Exception(message)

class PeripheralErrorException(val errorCode: Byte) :
    Exception("Peripheral error: 0x${errorCode.toInt().and(0xFF).toString(16).padStart(2, '0')}")

class ProtocolException(message: String) : Exception(message)

class BlerpcClient(context: Context) : GeneratedClient() {
    val transport = BleTransport(context)
    private var splitter: ContainerSplitter? = null
    private val assembler = ContainerAssembler()
    private var timeoutMs: Long = 100
    private var maxRequestPayloadSize: Int? = null
    private var maxResponsePayloadSize: Int? = null

    // Encryption state
    private var session: BlerpcCryptoSession? = null

    val mtu: Int get() = transport.mtu

    suspend fun scan(timeout: Long = 5000, serviceUuid: UUID? = SERVICE_UUID): List<ScannedDevice> {
        return transport.scan(timeout, serviceUuid)
    }

    suspend fun connect(device: ScannedDevice) {
        transport.connect(device)
        splitter = ContainerSplitter(mtu = transport.mtu)

        try {
            requestTimeout()
        } catch (_: TimeoutCancellationException) {
            Log.d(TAG, "Peripheral did not respond to timeout request, using default")
        }
        try {
            requestCapabilities()
        } catch (_: TimeoutCancellationException) {
            Log.d(TAG, "Peripheral did not respond to capabilities request")
        }
    }

    private suspend fun requestTimeout() {
        val s = splitter ?: throw IllegalStateException("Not connected")
        val tid = s.nextTransactionId()
        val req = makeTimeoutRequest(transactionId = tid)
        transport.write(req.serialize())
        val data = transport.readNotify(1000)
        val resp = Container.deserialize(data)
        if (resp.containerType == ContainerType.CONTROL &&
            resp.controlCmd == ControlCmd.TIMEOUT &&
            resp.payload.size == 2
        ) {
            val ms = ByteBuffer.wrap(resp.payload).order(ByteOrder.LITTLE_ENDIAN).short.toInt() and 0xFFFF
            timeoutMs = ms.toLong()
        } else {
            Log.w(TAG, "Unexpected timeout response: type=${resp.containerType}, cmd=${resp.controlCmd}, payload_len=${resp.payload.size}")
        }
    }

    private suspend fun requestCapabilities() {
        val s = splitter ?: throw IllegalStateException("Not connected")
        val tid = s.nextTransactionId()
        val req = makeCapabilitiesRequest(transactionId = tid)
        transport.write(req.serialize())
        val data = transport.readNotify(1000)
        val resp = Container.deserialize(data)
        if (resp.containerType == ContainerType.CONTROL &&
            resp.controlCmd == ControlCmd.CAPABILITIES &&
            resp.payload.size >= 6
        ) {
            val buf = ByteBuffer.wrap(resp.payload).order(ByteOrder.LITTLE_ENDIAN)
            val maxReq = buf.short.toInt() and 0xFFFF
            val maxResp = buf.short.toInt() and 0xFFFF
            val flags = buf.short.toInt() and 0xFFFF
            if (maxReq == 0 || maxResp == 0) {
                Log.w(TAG, "Peripheral reported zero capability: max_request=$maxReq, max_response=$maxResp")
            }
            maxRequestPayloadSize = maxReq
            maxResponsePayloadSize = maxResp
            Log.d(TAG, "Peripheral capabilities: max_request=$maxReq, max_response=$maxResp, flags=0x${flags.toString(16).padStart(4, '0')}")

            // Initiate key exchange if peripheral supports encryption
            if (flags and CAPABILITY_FLAG_ENCRYPTION_SUPPORTED != 0) {
                performKeyExchange()
            }
        } else {
            Log.w(TAG, "Unexpected capabilities response: type=${resp.containerType}, cmd=${resp.controlCmd}, payload_len=${resp.payload.size}")
        }
    }

    private suspend fun performKeyExchange() {
        val s = splitter ?: throw IllegalStateException("Not connected")
        val kx = CentralKeyExchange()

        // Step 1: Send central's ephemeral public key
        val step1 = kx.start()
        val tid1 = s.nextTransactionId()
        val req1 = makeKeyExchange(transactionId = tid1, payload = step1)
        transport.write(req1.serialize())

        // Step 2: Receive peripheral's response
        val data2 = transport.readNotify(2000)
        val resp2 = Container.deserialize(data2)
        if (resp2.containerType != ContainerType.CONTROL ||
            resp2.controlCmd != ControlCmd.KEY_EXCHANGE
        ) {
            Log.e(TAG, "Expected KEY_EXCHANGE step 2, got something else")
            return
        }

        val step3 = try {
            kx.processStep2(resp2.payload)
        } catch (e: IllegalArgumentException) {
            Log.e(TAG, "Key exchange step 2 failed: ${e.message}")
            return
        }

        // Step 3: Send encrypted confirmation
        val tid3 = s.nextTransactionId()
        val req3 = makeKeyExchange(transactionId = tid3, payload = step3)
        transport.write(req3.serialize())

        // Step 4: Receive peripheral's confirmation
        val data4 = transport.readNotify(2000)
        val resp4 = Container.deserialize(data4)
        if (resp4.containerType != ContainerType.CONTROL ||
            resp4.controlCmd != ControlCmd.KEY_EXCHANGE
        ) {
            Log.e(TAG, "Expected KEY_EXCHANGE step 4, got something else")
            return
        }

        session = try {
            kx.finish(resp4.payload)
        } catch (e: IllegalArgumentException) {
            Log.e(TAG, "Key exchange step 4 failed: ${e.message}")
            return
        }
        Log.i(TAG, "E2E encryption established")
    }

    private fun encryptPayload(payload: ByteArray): ByteArray {
        return session?.encrypt(payload) ?: payload
    }

    private fun decryptPayload(payload: ByteArray): ByteArray {
        return session?.decrypt(payload) ?: payload
    }

    override suspend fun call(cmdName: String, requestData: ByteArray): ByteArray {
        val s = splitter ?: throw IllegalStateException("Not connected")

        val cmd = CommandPacket(
            cmdType = CommandType.REQUEST,
            cmdName = cmdName,
            data = requestData
        )
        val payload = cmd.serialize()

        maxRequestPayloadSize?.let { limit ->
            if (payload.size > limit) throw PayloadTooLargeError(payload.size, limit)
        }

        // Encrypt if active, then split into containers and send
        val sendPayload = encryptPayload(payload)
        val containers = s.split(sendPayload)
        for (c in containers) {
            transport.write(c.serialize())
        }

        assembler.reset()
        var firstRead = true
        while (true) {
            // First read uses longer timeout (peripheral processing + BLE latency)
            val t = if (firstRead) maxOf(timeoutMs, 2000) else timeoutMs
            firstRead = false
            val notifyData = transport.readNotify(t)
            val container = Container.deserialize(notifyData)

            if (container.containerType == ContainerType.CONTROL) {
                if (container.controlCmd == ControlCmd.ERROR && container.payload.isNotEmpty()) {
                    val errorCode = container.payload[0]
                    if (errorCode == BLERPC_ERROR_RESPONSE_TOO_LARGE) {
                        throw ResponseTooLargeError("Response exceeds peripheral's max_response_payload_size")
                    }
                    throw PeripheralErrorException(errorCode)
                }
                continue
            }

            val result = assembler.feed(container)
            if (result != null) {
                // Decrypt if active
                val decrypted = decryptPayload(result)
                val resp = CommandPacket.deserialize(decrypted)
                if (resp.cmdType != CommandType.RESPONSE) {
                    throw ProtocolException("Expected response, got type=${resp.cmdType}")
                }
                if (resp.cmdName != cmdName) {
                    throw ProtocolException("Command name mismatch: expected '$cmdName', got '${resp.cmdName}'")
                }
                return resp.data
            }
        }
    }

    override suspend fun streamReceive(cmdName: String, requestData: ByteArray): List<ByteArray> {
        val s = splitter ?: throw IllegalStateException("Not connected")

        val cmd = CommandPacket(
            cmdType = CommandType.REQUEST,
            cmdName = cmdName,
            data = requestData
        )
        val payload = cmd.serialize()

        maxRequestPayloadSize?.let { limit ->
            if (payload.size > limit) throw PayloadTooLargeError(payload.size, limit)
        }

        // Encrypt if active, then split into containers and send
        val sendPayload = encryptPayload(payload)
        val containers = s.split(sendPayload)
        for (c in containers) {
            transport.write(c.serialize())
        }

        val results = mutableListOf<ByteArray>()
        assembler.reset()
        var firstRead = true
        while (true) {
            // First read uses longer timeout (peripheral processing time)
            val t = if (firstRead) maxOf(timeoutMs, 2000) else timeoutMs
            firstRead = false
            val notifyData = transport.readNotify(t)
            val container = Container.deserialize(notifyData)

            if (container.containerType == ContainerType.CONTROL) {
                if (container.controlCmd == ControlCmd.STREAM_END_P2C) {
                    break
                }
                if (container.controlCmd == ControlCmd.ERROR && container.payload.isNotEmpty()) {
                    val errorCode = container.payload[0]
                    if (errorCode == BLERPC_ERROR_RESPONSE_TOO_LARGE) {
                        throw ResponseTooLargeError("Response exceeds peripheral's max_response_payload_size")
                    }
                    throw PeripheralErrorException(errorCode)
                }
                continue
            }

            val result = assembler.feed(container)
            if (result != null) {
                // Decrypt each received response
                val decrypted = decryptPayload(result)
                val resp = CommandPacket.deserialize(decrypted)
                if (resp.cmdType != CommandType.RESPONSE) {
                    throw ProtocolException("Expected response, got type=${resp.cmdType}")
                }
                results.add(resp.data)
            }
        }
        return results
    }

    override suspend fun streamSend(
        cmdName: String,
        messages: List<ByteArray>,
        finalCmdName: String
    ): ByteArray {
        val s = splitter ?: throw IllegalStateException("Not connected")

        // Encrypt each message
        for (msgData in messages) {
            val cmd = CommandPacket(
                cmdType = CommandType.REQUEST,
                cmdName = cmdName,
                data = msgData
            )
            val payload = cmd.serialize()
            val sendPayload = encryptPayload(payload)
            val containers = s.split(sendPayload)
            for (c in containers) {
                transport.write(c.serialize())
            }
        }

        // Send STREAM_END_C2P
        val tid = s.nextTransactionId()
        val streamEnd = makeStreamEndC2P(transactionId = tid)
        transport.write(streamEnd.serialize())

        // Wait for final response
        assembler.reset()
        var firstRead = true
        while (true) {
            val t = if (firstRead) maxOf(timeoutMs, 2000) else timeoutMs
            firstRead = false
            val notifyData = transport.readNotify(t)
            val container = Container.deserialize(notifyData)

            if (container.containerType == ContainerType.CONTROL) {
                if (container.controlCmd == ControlCmd.ERROR && container.payload.isNotEmpty()) {
                    val errorCode = container.payload[0]
                    if (errorCode == BLERPC_ERROR_RESPONSE_TOO_LARGE) {
                        throw ResponseTooLargeError("Response exceeds peripheral's max_response_payload_size")
                    }
                    throw PeripheralErrorException(errorCode)
                }
                continue
            }

            val result = assembler.feed(container)
            if (result != null) {
                // Decrypt final response
                val decrypted = decryptPayload(result)
                val resp = CommandPacket.deserialize(decrypted)
                if (resp.cmdType != CommandType.RESPONSE) {
                    throw ProtocolException("Expected response, got type=${resp.cmdType}")
                }
                if (resp.cmdName != finalCmdName) {
                    throw ProtocolException("Command name mismatch: expected '$finalCmdName', got '${resp.cmdName}'")
                }
                return resp.data
            }
        }
    }

    suspend fun counterStreamAll(count: Int): List<Pair<Int, Int>> {
        val req = blerpc.Blerpc.CounterStreamRequest.newBuilder().setCount(count).build()
        val responses = streamReceive("counter_stream", req.toByteArray())
        return responses.map { data ->
            val resp = blerpc.Blerpc.CounterStreamResponse.parseFrom(data)
            Pair(resp.seq, resp.value)
        }
    }

    suspend fun counterUploadAll(count: Int): blerpc.Blerpc.CounterUploadResponse {
        val messages = (0 until count).map { i ->
            blerpc.Blerpc.CounterUploadRequest.newBuilder()
                .setSeq(i)
                .setValue(i * 10)
                .build()
                .toByteArray()
        }
        val respData = streamSend("counter_upload", messages, "counter_upload")
        return blerpc.Blerpc.CounterUploadResponse.parseFrom(respData)
    }

    fun disconnect() {
        transport.disconnect()
    }

    companion object {
        private const val TAG = "BlerpcClient"
    }
}
